### ISSUE TITLE
Usage of the parameter of candidate_ids and num_hard_negatives?

### ISSUE THREAD

**Issue by hpd123** [tensorflow/recommenders#330]
parameter num_hard_negatives: when i set it as an integer lower than batch_size, the loss would not slow down.
parameter candidate_ids : I do not know how to use it. 
 I will be grateful if anyone knows the usage of the above parameters and leave a message to me!

tfrs.tasks.Retrieval(
    loss: Optional[tf.keras.losses.Loss] = None,
    metrics: Optional[tfrs.metrics.FactorizedTopK] = None,
    temperature: Optional[float] = None,
    num_hard_negatives: Optional[int] = None,
    name: Optional[Text] = None
) -> None

call(
    query_embeddings: tf.Tensor,
    candidate_embeddings: tf.Tensor,
    sample_weight: Optional[tf.Tensor] = None,
    candidate_sampling_probability: Optional[tf.Tensor] = None,
    candidate_ids: Optional[tf.Tensor] = None,
    compute_metrics: bool = True
) -> tf.Tensor

**Comment by Andy1314Chen** [comment]
parameter `num_hard_negatives`: top-k largest logits are kept to calc the softmax loss
parameter `candidate_ids` to remove the same id when you use in-batch softmax

**Comment by ydennisy** [comment]
@Andy1314Chen thanks for the reply but could you elaborate on this a little more?
